File: src/rust/iced-x86-js/src/tests/info.test.js
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

const {
	Decoder, DecoderOptions, InstructionInfoFactory, InstructionInfoOptions, MemorySize,
	OpAccess, Register
} = require("iced-x86");

test("Instruction info factory", () => {
	// add [rdi+r12*8-5AA5EDCCh],esi
	const bytes = new Uint8Array([0x42, 0x01, 0xB4, 0xE7, 0x34, 0x12, 0x5A, 0xA5]);
	const decoder = new Decoder(64, bytes, DecoderOptions.None);
	const infoFactory = new InstructionInfoFactory();
	const instr = decoder.decode();

	const info1 = infoFactory.info(instr);
	expect(info1.op0Access).toBe(OpAccess.ReadWrite);
	expect(info1.op1Access).toBe(OpAccess.Read);
	expect(info1.op2Access).toBe(OpAccess.None);
	expect(info1.op3Access).toBe(OpAccess.None);
	expect(info1.op4Access).toBe(OpAccess.None);
	expect(info1.opAccess(0)).toBe(OpAccess.ReadWrite);
	expect(info1.opAccess(1)).toBe(OpAccess.Read);
	expect(info1.opAccess(2)).toBe(OpAccess.None);
	expect(info1.opAccess(3)).toBe(OpAccess.None);
	expect(info1.opAccess(4)).toBe(OpAccess.None);
	expect(instr.fpuTopIncrement).toBe(0);
	expect(instr.fpuCondWritesTop).toBe(false);
	expect(instr.fpuWritesTop).toBe(false);

	const regs1 = info1.usedRegisters();
	expect(regs1.length).toBe(3);
	expect(regs1[0].register).toBe(Register.RDI);
	expect(regs1[0].access).toBe(OpAccess.Read);
	expect(regs1[1].register).toBe(Register.R12);
	expect(regs1[1].access).toBe(OpAccess.Read);
	expect(regs1[2].register).toBe(Register.ESI);
	expect(regs1[2].access).toBe(OpAccess.Read);
	const mem1 = info1.usedMemory();
	expect(mem1.length).toBe(1);
	expect(mem1[0].segment).toBe(Register.DS);
	expect(mem1[0].base).toBe(Register.RDI);
	expect(mem1[0].index).toBe(Register.R12);
	expect(mem1[0].scale).toBe(8);
	expect(mem1[0].displacement).toBe(0xFFFFFFFFA55A1234n);
	expect(mem1[0].memorySize).toBe(MemorySize.UInt32);
	expect(mem1[0].access).toBe(OpAccess.ReadWrite);

	const info2 = infoFactory.infoOptions(instr, InstructionInfoOptions.NoMemoryUsage);
	const regs2 = info2.usedRegisters();
	expect(regs2.length).toBe(3);
	expect(regs2[0].register).toBe(Register.RDI);
	expect(regs2[0].access).toBe(OpAccess.Read);
	expect(regs2[1].register).toBe(Register.R12);
	expect(regs2[1].access).toBe(OpAccess.Read);
	expect(regs2[2].register).toBe(Register.ESI);
	expect(regs2[2].access).toBe(OpAccess.Read);
	const mem2 = info2.usedMemory();
	expect(mem2.length).toBe(0);

	const info3 = infoFactory.infoOptions(instr, InstructionInfoOptions.NoRegisterUsage);
	const regs3 = info3.usedRegisters();
	expect(regs3.length).toBe(0);
	const mem3 = info3.usedMemory();
	expect(mem3.length).toBe(1);
	expect(mem3[0].segment).toBe(Register.DS);
	expect(mem3[0].base).toBe(Register.RDI);
	expect(mem3[0].index).toBe(Register.R12);
	expect(mem3[0].scale).toBe(8);
	expect(mem3[0].displacement).toBe(0xFFFFFFFFA55A1234n);
	expect(mem3[0].memorySize).toBe(MemorySize.UInt32);
	expect(mem3[0].access).toBe(OpAccess.ReadWrite);

	const info4 = infoFactory.infoOptions(instr, InstructionInfoOptions.NoMemoryUsage | InstructionInfoOptions.NoRegisterUsage);
	const regs4 = info4.usedRegisters();
	expect(regs4.length).toBe(0);
	const mem4 = info4.usedMemory();
	expect(mem4.length).toBe(0);

	decoder.free();
	infoFactory.free();
	instr.free();
	info1.free();
	info2.free();
	info3.free();
	info4.free();
	regs1.forEach(a => a.free());
	regs2.forEach(a => a.free());
	regs3.forEach(a => a.free());
	regs4.forEach(a => a.free());
	mem1.forEach(a => a.free());
	mem2.forEach(a => a.free());
	mem3.forEach(a => a.free());
	mem4.forEach(a => a.free());
});

File: src/rust/iced-x86-py/tests/info_test.py
# SPDX-License-Identifier: MIT
# Copyright (C) 2018-present iced project and contributors

import copy
import pytest
from iced_x86 import *
from typing import Callable

def test_info() -> None:
	instr = Decoder(64, b"\xC4\xE3\x49\x48\x10\x41").decode()
	factory = InstructionInfoFactory()
	info = factory.info(instr)

	assert info.op0_access == OpAccess.WRITE
	assert info.op1_access == OpAccess.READ
	assert info.op2_access == OpAccess.READ
	assert info.op3_access == OpAccess.READ
	assert info.op4_access == OpAccess.READ
	assert info.op_access(0) == info.op0_access
	assert info.op_access(1) == info.op1_access
	assert info.op_access(2) == info.op2_access
	assert info.op_access(3) == info.op3_access
	assert info.op_access(4) == info.op4_access

	mem_list = info.used_memory()
	assert isinstance(mem_list, list)
	assert len(mem_list) == 1
	mem = mem_list[0]
	assert mem.segment == Register.DS
	assert mem.base == Register.RAX
	assert mem.index == Register.NONE
	assert mem.scale == 1
	assert mem.displacement == 0
	assert mem.displacement_i64 == 0
	assert mem.memory_size == MemorySize.PACKED128_FLOAT32
	assert mem.access == OpAccess.READ
	assert mem.address_size == CodeSize.CODE64
	assert mem.vsib_size == 0

	reg_list = info.used_registers()
	assert isinstance(reg_list, list)
	assert len(reg_list) == 4
	assert (reg_list[0].register, reg_list[0].access) == (Register.ZMM2, OpAccess.WRITE)
	assert (reg_list[1].register, reg_list[1].access) == (Register.XMM6, OpAccess.READ)
	assert (reg_list[2].register, reg_list[2].access) == (Register.RAX, OpAccess.READ)
	assert (reg_list[3].register, reg_list[3].access) == (Register.XMM4, OpAccess.READ)

@pytest.mark.parametrize("copy_value", [
	lambda instr: copy.copy(instr),
	lambda instr: copy.deepcopy(instr),
	lambda instr: instr.copy(),
])
def test_copy_deepcopy_mcopy_used_mem(copy_value: Callable[[UsedMemory], UsedMemory]) -> None:
	instr = Decoder(64, b"\xC4\xE3\x49\x48\x10\x41").decode()
	factory = InstructionInfoFactory()
	info = factory.info(instr)

	mem = info.used_memory()[0]
	mem2 = copy_value(mem)
	assert mem is not mem2
	assert id(mem) != id(mem2)
	assert mem == mem2
	assert not (mem != mem2)
	assert hash(mem) == hash(mem2)

@pytest.mark.parametrize("copy_value", [
	lambda instr: copy.copy(instr),
	lambda instr: copy.deepcopy(instr),
	lambda instr: instr.copy(),
])
def test_copy_deepcopy_mcopy_used_reg(copy_value: Callable[[UsedRegister], UsedRegister]) -> None:
	instr = Decoder(64, b"\xC4\xE3\x49\x48\x10\x41").decode()
	factory = InstructionInfoFactory()
	info = factory.info(instr)

	reg = info.used_registers()[0]
	reg2 = copy_value(reg)
	assert reg is not reg2
	assert id(reg) != id(reg2)
	assert reg == reg2
	assert not (reg != reg2)
	assert hash(reg) == hash(reg2)

def test_op_access_raise() -> None:
	instr = Decoder(64, b"\xC4\xE3\x49\x48\x10\x41").decode()
	factory = InstructionInfoFactory()
	info = factory.info(instr)

	info.op_access(4)
	with pytest.raises(ValueError):
		info.op_access(123)
